### PR TITLE
impl(pubsub): add tracing for exactly once lease management

### PR DESCRIPTION
**Pull request by alevenberg**
https://github.com/googleapis/google-cloud-cpp/issues/13726

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13952)

### CODE REVIEW

**Comment by alevenberg** [comment]
Ready for review

File: google/cloud/pubsub/internal/extend_leases_with_retry.cc [hunk 1]
```diff
@@ -56,6 +57,32 @@ google::pubsub::v1::ModifyAckDeadlineRequest UpdateRequest(
   return request;
 }
 
+class ExtendLeasesWithRetryImpl {
+ public:
+  virtual ~ExtendLeasesWithRetryImpl() = default;
+  virtual future<Status> MakeRpc(
```
Comment (devbww) [inline google/cloud/pubsub/internal/extend_leases_with_retry.cc:63]: Perhaps `MakeRpc()` might be better as `AsyncModifyAckDeadline()`?

File: google/cloud/pubsub/internal/extend_leases_with_retry.cc [hunk 2]
```diff
@@ -56,6 +57,32 @@ google::pubsub::v1::ModifyAckDeadlineRequest UpdateRequest(
...
+  future<Status> MakeRpc(
+      CompletionQueue& cq,
+      google::pubsub::v1::ModifyAckDeadlineRequest const& request) override {
+    auto context = std::make_shared<grpc::ClientContext>();
+    auto options = google::cloud::internal::MakeImmutableOptions({});
+    return stub_->AsyncModifyAckDeadline(cq, std::move(context),
+                                         std::move(options), request);
+  }
+
+  std::shared_ptr<SubscriberStub> stub_;
```
Comment (devbww) [inline google/cloud/pubsub/internal/extend_leases_with_retry.cc:83]: `private`?

File: google/cloud/pubsub/internal/extend_leases_with_retry.cc [hunk 3]
```diff
@@ -151,21 +174,75 @@ class ExtendLeasesWithRetryHandle
...
+          auto result = f.get();
+          for (auto const& ack_id : request.ack_ids()) {
+            cb->ModackEnd(ack_id);
+          }
+          cb->EndModackSpan(modack_span);
+          return result;
+        });
+  }
+
+  std::shared_ptr<ExtendLeasesWithRetryImpl> child_;
```
Comment (devbww) [inline google/cloud/pubsub/internal/extend_leases_with_retry.cc:210]: `private`?

File: google/cloud/pubsub/internal/extend_leases_with_retry_test.cc [hunk 4]
```diff
@@ -112,56 +130,148 @@ TEST(ExtendLeasesWithRetry, FailurePermanentError) {
...
+  request.add_ack_ids("test-002");
+
+  auto result =
+      ExtendLeasesWithRetry(mock, CompletionQueue(mock_cq), request,
+                            mock_batch_callback, /*enable_otel=*/true);
+
+  EXPECT_STATUS_OK(result.get());
+}
+
+TEST(ExtendLeasesWithRetry, SuccessWithOtelEnabled) {
```
Comment (devbww) [inline google/cloud/pubsub/internal/extend_leases_with_retry_test.cc:234]: I couldn't see how this test was different (apart from some whitespace differences) to the previous one.

File: google/cloud/pubsub/internal/extend_leases_with_retry.cc [hunk 5]
```diff
@@ -56,6 +57,32 @@ google::pubsub::v1::ModifyAckDeadlineRequest UpdateRequest(
   return request;
 }
 
+class ExtendLeasesWithRetryImpl {
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/internal/extend_leases_with_retry.cc:60]: It is an Impl for `ExtendLeasesWithRetryHandle`, but I see your point. Renamed to 	`ExtendLeases`.

File: google/cloud/pubsub/internal/extend_leases_with_retry.cc (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/extend_leases_with_retry.cc:63]: Done

File: google/cloud/pubsub/internal/extend_leases_with_retry.cc (hunk 3 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/extend_leases_with_retry.cc:210]: Done

File: google/cloud/pubsub/internal/extend_leases_with_retry.cc (hunk 2 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/extend_leases_with_retry.cc:83]: Done

File: google/cloud/pubsub/internal/extend_leases_with_retry_test.cc (hunk 4 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/extend_leases_with_retry_test.cc:234]: You're right I think they are the same. I think it was left over from a merge. Removed.

**Review by devbww — APPROVED** [review]

**Review by dbolduc — COMMENTED** [review]
feel free to ignore

**Pull request merged**